src/client/parse_func.c

```diff
@@ -566,61 +568,48 @@ xddfunc_datapattern(xdd_plan_t *planp, int32_t argc, char *argv[], uint32_t flag
 					(char *)argv[args+2]);
 				return(0);
 			}
-			pattern_length = stat_buf.st_size;
+		
 		}
 		if (tdp) { /* set option for specific target */
 			if (strcmp(pattern_type, "file") == 0) {
 				tdp->td_dpp->data_pattern_options |= DP_FILE_PATTERN;
 			} else {// wholefile
 				tdp->td_dpp->data_pattern_options |= DP_WHOLEFILE_PATTERN;
+				if (!xdd_datapattern_wholefile_enough_ram(tdp, argv[args+2])) {
+					return(0);
+				}
 			}
-			tdp->td_dpp->data_pattern_filename = (char *)argv[args+2];
-			tdp->td_dpp->data_pattern_length = pattern_length;
-			if (!xdd_datapattern_wholefile_enough_ram(tdp)) {
-				fprintf(xgp->errout, "%s: file %s can not be loaded into memory because 50%% of RAM is "
-					"currently unavailable\n",
-					xgp->progname,
-					(char *)argv[args+2]);
-				return(0);
-			}
-			tdp->dpp_fd = open(tdp->td_dpp->data_pattern_filename, O_RDONLY);
-			if (tdp->dpp_fd < 0) {
-				fprintf(xgp->errout, "%s: could not open %s\n", xgp->progname, tdp->td_dpp->data_pattern_filename);
-				return(0);
+			tdp->td_dpp->data_pattern_length = stat_buf.st_size;
+			if(xdd_set_datapattern_from_filename(tdp, argv[args+2])) {
+				return (0);
 			}
-			tdp->td_dpp->data_pattern = (unsigned char *)malloc(sizeof(unsigned char) * (tdp->td_dpp->data_pattern_length + 1));
-			memset(tdp->td_dpp->data_pattern, '\0', sizeof(unsigned char) * (tdp->td_dpp->data_pattern_length + 1));
-			pread(tdp->dpp_fd, tdp->td_dpp->data_pattern, tdp->td_dpp->data_pattern_length, 0);
 		} else {// Put this option into all Targets
 			if (flags & XDD_PARSE_PHASE2) {
 				tdp = planp->target_datap[0];
+				tdp->td_dpp->data_pattern_length = stat_buf.st_size;
 				i = 0;
-				while (tdp) {
-					if (strcmp(pattern_type, "file") == 0) {
-						tdp->td_dpp->data_pattern_options |= DP_FILE_PATTERN;
-					} else {//wholefile
-						tdp->td_dpp->data_pattern_options |= DP_WHOLEFILE_PATTERN;
-					}
-					tdp->td_dpp->data_pattern_filename = (char *)argv[args+2];
-					tdp->td_dpp->data_pattern_length = pattern_length;
-					if (!xdd_datapattern_wholefile_enough_ram(tdp)) {
-						fprintf(xgp->errout, "%s: the file %s is larger than 50%% of total RAM\n",
-							xgp->progname,
-							(char *)argv[args+2]);
+				dp_fd = open(argv[args+2], O_RDONLY);
+				if (dp_fd < 0) {
+					fprintf(xgp->errout, "%s: could not open %s\n", xgp->progname, argv[args+2]);
+					return (0);
+				}
+				if (strcmp(pattern_type, "file") == 0) {
+					tdp->td_dpp->data_pattern_options |= DP_FILE_PATTERN;
+				} else {//wholefile
+					tdp->td_dpp->data_pattern_options |= DP_WHOLEFILE_PATTERN;
+					if (!xdd_datapattern_wholefile_enough_ram(tdp, argv[args+2])) {
 						return(0);
 					}
-					tdp->dpp_fd = open(tdp->td_dpp->data_pattern_filename, O_RDONLY);
-					if (tdp->dpp_fd < 0) {
-						fprintf(xgp->errout, "%s: could not open %s\n", xgp->progname, tdp->td_dpp->data_pattern_filename);
-						return(0);
+				}
+				while (tdp) {
```

Comment:
does this while need to be moved to line 588? `tdp->td_dpp->data_pattern_options` is only set for the first `tdp`

Reply:
So the other part is used to only set the data pattern from a file on a specific target. That is why the while loop is not part of that code branch.